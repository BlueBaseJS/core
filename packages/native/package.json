{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
