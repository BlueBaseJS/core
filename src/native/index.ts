export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
