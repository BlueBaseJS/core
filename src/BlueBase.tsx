import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}