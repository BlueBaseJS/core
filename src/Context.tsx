import { BlueBase } from './BlueBase';
import { createContext } from 'react';

export const BlueBaseContext: React.Context<BlueBase> = createContext(undefined as any);
export const BlueBaseProvider = BlueBaseContext.Provider;
export const BlueBaseConsumer = BlueBaseContext.Consumer;
