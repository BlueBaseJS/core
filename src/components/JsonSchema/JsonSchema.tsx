import { JsonComponentNode, JsonSchemaParser } from '../../lib/json-schema-parser';
import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { BlueBaseFilter } from '../../index';
import { MaybeArray } from '../../utils';
import React from 'react';

export interface JsonSchemaProps {

	/** JSON Schema. */
	schema: MaybeArray<JsonComponentNode>;

	/** Event name to filter this schema. If this is not provided, the schema is not filtered. */
	filter?: string;

	/** Arguments for the filter. */
	args?: { [key: string]: any };

  /**
   * Used to locate this view in end-to-end tests.
   */
	testID?: string,
}

const getComponent = (BB: BlueBase) => {
	return ({ component }: JsonComponentNode) => {

		return BB.Components.has(String(component)) ? BB.Components.resolve(String(component)) : null;
	};
};

/**
 * # üç± JsonSchema
 *
 * Renders a Component based on JSON schema. This allows developers to create dynamic
 * layouts in their apps, and even save the schema to databases.
 *
 * Moreover, it also makes that schema filter-able. So that any plugin can modify that schema
 * on runtime.
 *
 * ## Usage:
 * ```jsx
 * <JsonSchema
 * 	filter="content-filter"
 * 	args={{ style: { color: 'blue' } }}
 *  schema={{
 * 	 component: 'Text',
 * 	 props: {
 * 		 style: {
 * 			 color: 'red'
 * 		 }
 * 	 },
 * 	 text: 'This is the page content.',
 *  }
 * } />
 * ```
 */
export class JsonSchema extends React.PureComponent<JsonSchemaProps> {

	static contextType = BlueBaseContext;

	render() {

		const BB: BlueBase = this.context;

		const { filter, schema, args } = this.props;
		const parser = new JsonSchemaParser(getComponent(BB));

		// There's no filter, we don't need to do complex async handling
		if (!filter) {
			return parser.parseSchema(schema);
		}

		const children = (loadedSchema: MaybeArray<JsonComponentNode>) => parser.parseSchema(loadedSchema);

		return (
			<BlueBaseFilter
				filter={filter}
				value={schema}
				args={args}
				children={children}
			/>
		);
	}
}
