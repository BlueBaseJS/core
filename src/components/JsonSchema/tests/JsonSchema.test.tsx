import * as Native from '../../../native';
import { BlueBase } from '../../../BlueBase';
import { BlueBaseProvider } from '../../../Context';
import { JsonSchema } from '../JsonSchema';
import React from 'react';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

const BB = new BlueBase();
BB.Components.register('Text', Native.Text);

describe('JsonSchema', () => {
	const JsonSchemaWithProvider = (props: any) => (
		<BlueBaseProvider value={BB}>
			<JsonSchema {...props}/>
		</BlueBaseProvider>
	);
	test(`Snapshot ComponentState component`, () => {
		const component = TestRenderer.create(
			<JsonSchemaWithProvider schema={{
				component: 'Text',
				props: {
					style: {
						color: 'red'
					}
				},
				text: 'This componenet is generated through JsonSchema Component',
			}}/>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
