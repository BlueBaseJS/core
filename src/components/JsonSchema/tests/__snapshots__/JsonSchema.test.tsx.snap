// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchema Snapshot ComponentState component 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
>
  This componenet is generated through JsonSchema Component
</Text>
`;
